build_modules/lib/src/meta_module.dart

```diff
@@ -0,0 +1,433 @@
+  /// internal srcs).
+  final Set<AssetId> externalDeps;
+
+  /// Order in which this node was discovered.
+  int discoveryIndex;
+
+  /// Lowest discoveryIndex for any node this is connected to.
+  int lowestLinkedDiscoveryIndex;
+
+  _AssetNode(this.id, this.internalDeps, this.parts, this.externalDeps);
+
+  /// Creates an [_AssetNode] for [id] given a parsed [CompilationUnit] and some
+  /// [internalSrcs] which represent other assets that may become part of the
+  /// same module.
+  factory _AssetNode.forParsedUnit(
+      AssetId id, CompilationUnit parsed, Set<AssetId> internalSrcs) {
+    var externalDeps = new Set<AssetId>();
+    var internalDeps = new Set<AssetId>();
+    var parts = new Set<AssetId>();
+    for (var directive in parsed.directives) {
+      if (directive is! UriBasedDirective) continue;
+      var linkedId = _importUriToAssetId(
+          id, (directive as UriBasedDirective).uri.stringValue);
+      if (linkedId == null) continue;
+      if (directive is PartDirective) {
+        if (!internalSrcs.contains(linkedId)) {
+          throw new StateError(
+              'Referenced part file $linkedId from $id which is not in the '
+              'same package');
+        }
+        parts.add(linkedId);
+      } else if (internalSrcs.contains(linkedId)) {
+        internalDeps.add(linkedId);
+      } else {
+        externalDeps.add(linkedId);
+      }
+    }
+    return new _AssetNode(id, internalDeps, parts, externalDeps);
+  }
+}
+
+/// Creates simple modules based strictly off of [connectedComponents].
+///
+/// This creates more modules than we want, but we collapse them later on.
+Map<AssetId, Module> _createModulesFromComponents(
+    Iterable<Set<_AssetNode>> connectedComponents) {
+  var modules = <AssetId, Module>{};
+  for (var componentNodes in connectedComponents) {
+    // Name components based on first alphabetically sorted node, preferring
+    // public srcs (not under lib/src).
+    var sortedNodes = componentNodes.toList()
+      ..sort((a, b) => a.id.path.compareTo(b.id.path));
+    var primaryNode = sortedNodes.firstWhere(
+        (node) => !node.id.path.startsWith('lib/src/'),
+        orElse: () => sortedNodes.first);
+    // Expand to include all the part files of each node, these aren't
+    // included as individual `_AssetNodes`s in `connectedComponents`.
+    var allAssetIds =
+        componentNodes.expand((node) => [node.id]..addAll(node.parts)).toSet();
+    var allDepIds = new Set<AssetId>();
+    for (var node in componentNodes) {
+      allDepIds.addAll(node.externalDeps);
+      for (var id in node.internalDeps) {
+        if (allAssetIds.contains(id)) continue;
+        allDepIds.add(id);
+      }
+    }
+    var module = new Module(primaryNode.id, allAssetIds, allDepIds);
+    modules[module.primarySource] = module;
+  }
+  return modules;
+}
+
+/// Computes the strongly connected components reachable from entrypoints.
+List<Set<_AssetNode>> _stronglyConnectedComponents(
```

Comment:
Note that we have a canonical strongly connected components implementation now - we should migrate this to use that. I think we still want this class.